pkg/filter/wasm/wasm.go

```diff
@@ -0,0 +1,317 @@
+	return io.ReadAll(resp.Body)
+}
+
+func isURL(str string) bool {
+	// we only check the first a few bytes as str could be
+	// the LONG base64 encoded wasm code
+	for _, p := range []string{"http://", "https://"} {
+		if len(str) > len(p) {
+			s := strings.ToLower(str[:len(p)])
+			if strings.HasPrefix(s, p) {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+func (f *WasmFilter) readWasmCode() ([]byte, error) {
+	if isURL(f.spec.Code) {
+		return readWasmCodeFromURL(f.spec.Code)
+	}
+	if _, e := os.Stat(f.spec.Code); e == nil {
+		return readWasmCodeFromFile(f.spec.Code)
+	}
+	return base64.StdEncoding.DecodeString(f.spec.Code)
+}
+
+func (f *WasmFilter) loadWasmCode() error {
+	code, e := f.readWasmCode()
+	if e != nil {
+		logger.Errorf("failed to load Wasm code: %v", e)
+		return e
+	}
+
+	if len(f.code) > 0 && bytes.Equal(f.code, code) {
+		return nil
+	}
+
+	p, e := NewWasmVMPool(f.spec.MaxConcurrency, code)
+	if e != nil {
+		logger.Errorf("failed to create Wasm VM pool: %v", e)
+		return e
+	}
+	f.code = code
+
+	f.vmPool.Store(p)
+	return nil
+}
+
+func (f *WasmFilter) watchWasmCode() {
+	var (
+		chWasm <-chan *string
+		syncer *cluster.Syncer
+		err    error
+	)
+
+	for {
+		c := f.pipeSpec.Super().Cluster()
+		syncer, err = c.Syncer(time.Minute)
+		if err == nil {
+			chWasm, err = syncer.Sync(c.Layout().WasmCodeEvent())
+			if err == nil {
+				break
+			}
+		}
+		logger.Errorf("failed to watch Wasm code event: %v", err)
+		select {
+		case <-time.After(10 * time.Second):
+		case <-f.chStop:
+			return
+		}
+	}
+
+	for {
+		select {
+		case <-chWasm:
+			err = f.loadWasmCode()
+
+		case <-time.After(30 * time.Second):
+			if err != nil || len(f.code) == 0 {
+				err = f.loadWasmCode()
+			}
+
+		case <-f.chStop:
+			return
+		}
+	}
+}
+
+func (f *WasmFilter) reload(pipeSpec *httppipeline.FilterSpec) {
+	f.pipeSpec = pipeSpec
+	f.spec = pipeSpec.FilterSpec().(*Spec)
+
+	f.spec.timeout, _ = time.ParseDuration(f.spec.Timeout)
+	f.chStop = make(chan struct{})
+
+	f.loadWasmCode()
+	go f.watchWasmCode()
+}
+
+// Init initializes WasmFilter.
+func (f *WasmFilter) Init(pipeSpec *httppipeline.FilterSpec) {
+	f.reload(pipeSpec)
+}
+
+// Inherit inherits previous generation of WasmFilter.
+func (f *WasmFilter) Inherit(pipeSpec *httppipeline.FilterSpec, previousGeneration httppipeline.Filter) {
+	previousGeneration.Close()
+	f.reload(pipeSpec)
+}
+
+// Handle handles HTTP request
+func (f *WasmFilter) Handle(ctx context.HTTPContext) string {
+	result := f.handle(ctx)
+	return ctx.CallNextHandler(result)
+}
+
+func (f *WasmFilter) handle(ctx context.HTTPContext) (result string) {
+	// we must save the pool to a local variable for later use as it will be
+	// replaced when updating the Wasm code
+	var pool *WasmVMPool
+	if p := f.vmPool.Load(); p == nil {
+		ctx.AddTag("Wasm VM pool is not initialized")
+		return resultOutOfVM
+	} else {
+		pool = p.(*WasmVMPool)
+	}
+
+	// get a free WASM VM and attach the ctx to it
+	vm := pool.Get()
```

Comment:
`pool.Get()` may create a new VM by calling `newWasmVM` which will invoke `panic` if something goes wrong. So we may also need to use `recover` here if we want to avoid the whole request cllapsed.

Reply:
where the `panic` come from?
if it comes from `importHostFuncs`, then that's fine because it should never happen and must be fixed at the development phase.

Reply:
Yes, such as `writeStringToWasm`'s `memory allocate panic`. Since we had discussed the background clearly, it's only a registration behavior instead of actual code execution, I am Ok with it.